File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/28．单位驻留进度.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local GetHandleId = jass.GetHandleId
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
local function _____53D6_5355_4F4DID(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    return GetHandleId(unit) or 0
end
____exports["创建单位驻留进度"] = function(_____540D_79F0, _____5141_8BB8_4F4D_79FB_8DDD_79BB)
    local _____8BB0_5F55_8868 = {}
    local _____6709_6548_4F4D_79FB_8DDD_79BB = _____5141_8BB8_4F4D_79FB_8DDD_79BB > 0 and _____5141_8BB8_4F4D_79FB_8DDD_79BB or 0
    local _____4F4D_79FB_8DDD_79BB_5E73_65B9 = _____6709_6548_4F4D_79FB_8DDD_79BB * _____6709_6548_4F4D_79FB_8DDD_79BB
    return {
        ["名称"] = _____540D_79F0,
        ["采样"] = function(unit, _____589E_91CF)
            if _____589E_91CF == nil then
                _____589E_91CF = 1
            end
            local id = _____53D6_5355_4F4DID(unit)
            if id == 0 then
                return 0
            end
            local x = GetUnitX(unit)
            local y = GetUnitY(unit)
            local _____8BB0_5F55 = _____8BB0_5F55_8868[id]
            if _____8BB0_5F55 == nil or (x - _____8BB0_5F55.X) * (x - _____8BB0_5F55.X) + (y - _____8BB0_5F55.Y) * (y - _____8BB0_5F55.Y) > _____4F4D_79FB_8DDD_79BB_5E73_65B9 then
                _____8BB0_5F55 = {X = x, Y = y, ["进度"] = 0}
                _____8BB0_5F55_8868[id] = _____8BB0_5F55
            end
            _____8BB0_5F55["进度"] = _____8BB0_5F55["进度"] + _____589E_91CF
            return _____8BB0_5F55["进度"]
        end,
        ["读取"] = function(unit)
            local ____opt_0 = _____8BB0_5F55_8868[_____53D6_5355_4F4DID(unit)]
            return ____opt_0 and ____opt_0["进度"] or 0
        end,
        ["清空"] = function(unit)
            if unit == nil then
                _____8BB0_5F55_8868 = {}
                return
            end
            __TS__Delete(
                _____8BB0_5F55_8868,
                _____53D6_5355_4F4DID(unit)
            )
        end
    }
end
return ____exports

File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/09．装备通用机制/28．单位驻留进度.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const GetHandleId = jass.GetHandleId as (handle: any) => number;
const GetUnitX = jass.GetUnitX as (unit: any) => number;
const GetUnitY = jass.GetUnitY as (unit: any) => number;

interface 单位驻留记录 {
  X: number;
  Y: number;
  进度: number;
}

export interface 单位驻留进度控制器 {
  readonly 名称: string;
  采样(this: void, unit: any, 增量?: number): number;
  读取(this: void, unit: any): number;
  清空(this: void, unit?: any): void;
}

function 取单位ID(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  return GetHandleId(unit) || 0;
}

export function 创建单位驻留进度(this: void, 名称: string, 允许位移距离: number): 单位驻留进度控制器 {
  let 记录表: Record<number, 单位驻留记录 | undefined> = {};
  const 有效位移距离 = 允许位移距离 > 0 ? 允许位移距离 : 0;
  const 位移距离平方 = 有效位移距离 * 有效位移距离;

  return {
    名称,
    采样: function 采样(this: void, unit: any, 增量: number = 1): number {
      const id = 取单位ID(unit);
      if (id === 0) return 0;
      const x = GetUnitX(unit);
      const y = GetUnitY(unit);
      let 记录 = 记录表[id];
      if (记录 == null || (x - 记录.X) * (x - 记录.X) + (y - 记录.Y) * (y - 记录.Y) > 位移距离平方) {
        记录 = { X: x, Y: y, 进度: 0 };
        记录表[id] = 记录;
      }
      记录.进度 += 增量;
      return 记录.进度;
    },
    读取: function 读取(this: void, unit: any): number {
      return 记录表[取单位ID(unit)]?.进度 ?? 0;
    },
    清空: function 清空(this: void, unit?: any): void {
      if (unit == null) {
        记录表 = {};
        return;
      }
      delete 记录表[取单位ID(unit)];
    },
  };
}

export {};
